### PR TITLE
[ODS-4668] Code-gen for extension plugin creation

### PR DESCRIPTION
**Pull request by semalaiappan**
Different usecase as below
Invoke-CodeGen -Engine SQLServer -IncludePlugins
Invoke-CodeGen -Engine SQLServer -IncludePlugins -ExtensionPaths @('C:\OSS-Workspace\Semalai_EndUser','C:\OSS-Workspace\EdFi.Ods.Extensions.TPDM')
Invoke-CodeGen -Engine SQLServer -ExtensionPaths @('C:\OSS-Workspace\Semalai_EndUser','C:\OSS-Workspace\EdFi.Ods.Extensions.TPDM')
Invoke-CodeGen -Engine SQLServer
Invoke-CodeGen -Engine SQLServer -IncludePlugins -ExtensionPaths @('C:\OSS-Workspace\Ed-Fi-Extensions\Extensions')

### CODE REVIEW

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs [hunk 1]
```diff
@@ -32,7 +33,18 @@ protected override void Arrange()
...
                     .Returns(false);
 
-                _domainModelDefinitionProvidersProvider = new DomainModelDefinitionProvidersProvider(_codeRepositoryProvider, _includeExtensionsProvider);
+                A.CallTo(() => _codeRepositoryProvider.GetResolvedCodeRepositoryByName(A<string>._, A<string>._))
+                    .Returns(codeRepositoryHelper[CodeRepositoryConventions.ExtensionsFolderName]);
+
+                var extensionsPath = _codeRepositoryProvider.GetResolvedCodeRepositoryByName(
+                    CodeRepositoryConventions.ExtensionsFolderName,
+                    "Extensions");
+                A.CallTo(() => _extensionLocationPluginsProvider.GetExtensionLocationPlugins())
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs:42]: Add a blank line

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs [hunk 2]
```diff
@@ -32,7 +33,18 @@ protected override void Arrange()
...
+                    .Returns(new[]
+                    {
+                        extensionsPath
+                    });
+                _domainModelDefinitionProvidersProvider = new DomainModelDefinitionProvidersProvider(_codeRepositoryProvider, _extensionLocationPluginsProvider, _includeExtensionsProvider);
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs:47]: add a blank line

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs [hunk 3]
```diff
@@ -21,5 +22,8 @@ public class Options
 
         [Option('i', "includePlugins", Required = false, HelpText = "Flag to determine if plugin assemblies are included", Default = false)]
         public bool IncludePlugins { get; set; }
+
+        [Option('j', "extensionLocationPaths", Required = false, HelpText = "Array of paths for the Extension location to determine if plugin assemblies are included")]
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs:26]: is 'j' the right parameter? what other options do we have?

Comment (jamessmckay) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs:26]: should the long parameter be `extensionsPaths`

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs [hunk 4]
```diff
@@ -30,12 +29,13 @@ public class AssemblyDataProvider : IAssemblyDataProvider
...
         private readonly IIncludePluginsProvider _includePluginsProvider;
-        
+        private readonly IExtensionLocationPluginsProvider _extensionLocationPluginsProviderProvider;
+
         public AssemblyDataProvider(
             ICodeRepositoryProvider codeRepositoryProvider,
             IJsonFileProvider jsonFileProvider,
             IDomainModelDefinitionsProviderProvider domainModelDefinitionsProviderProvider,
-            IIncludePluginsProvider includePluginsProvider)
+            IIncludePluginsProvider includePluginsProvider, IExtensionLocationPluginsProvider extensionLocationPluginsProviderProvider)
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:38]: fix formatting please

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs [hunk 5]
```diff
@@ -70,7 +72,26 @@ public IEnumerable<AssemblyData> Get()
                 .Select(Create)
                 .ToList();
 
-            if (_includePluginsProvider.IncludePlugins())
+            var extensionLocationPaths = _extensionLocationPluginsProviderProvider.GetExtensionLocationPlugins();
+
+            if (extensionLocationPaths != null && extensionLocationPaths.Any())
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:77]: This if statement is redundant since you are doing a foreach loop

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs [hunk 6]
```diff
@@ -70,7 +72,26 @@ public IEnumerable<AssemblyData> Get()
...
+                                    extensionLocationPath,
+                                    AssemblyMetadataSearchString,
+                                    SearchOption.AllDirectories)
+                                .Select(Create)
+                                .ToList()
+                        );
+                    }
+                }
+            }
+            else if (_includePluginsProvider.IncludePlugins())
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:94]: This statement is an independent event

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 7]
```diff
@@ -57,33 +60,47 @@ public IEnumerable<IDomainModelDefinitionsProvider> DomainModelDefinitionProvide
 
         private Dictionary<string, IDomainModelDefinitionsProvider> CreateDomainModelDefinitionsByPath()
         {
-            DirectoryInfo[] directoriesToEvaluate;
-            
+            DirectoryInfo[] directoriesToEvaluate = new DirectoryInfo[3];
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:63]: why this change?

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 8]
```diff
@@ -57,33 +60,47 @@ public IEnumerable<IDomainModelDefinitionsProvider> DomainModelDefinitionProvide
...
                 CodeRepositoryConventions.EdFiOdsFolderName);
 
-            if (_includePluginsProvider.IncludePlugins() && Directory.Exists(_extensionsPath))
+            directoriesToEvaluate = GetProjectDirectoriesToEvaluate(edFiOdsImplementationApplicationPath)
+                .Concat(GetProjectDirectoriesToEvaluate(edFiOdsApplicationPath))
+                .ToArray();
+
+            var extensionLocationPaths = _extensionLocationPluginsProviderProvider.GetExtensionLocationPlugins();
+
+            if (extensionLocationPaths != null && extensionLocationPaths.Any())
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:80]: this statement is redundant with the for each statement

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 9]
```diff
@@ -57,33 +60,47 @@ public IEnumerable<IDomainModelDefinitionsProvider> DomainModelDefinitionProvide
...
+                    }
+                    else
+                    {
+                        throw new Exception(
+                            $"Unable to find extension Location project path  at location {extensionLocationPath}.");
+                    }
+                }
             }
-            else
+            else if (_includePluginsProvider.IncludePlugins() && Directory.Exists(_extensionsPath))
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:96]: this is an independent statement (ie. the else if should only be an if)

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 10]
```diff
@@ -121,6 +138,7 @@ DirectoryInfo[] GetProjectDirectoriesToEvaluate(string basePath)
 
                 return new DirectoryInfo[0];
             }
+
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:141]: remove blank line

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        public ExtensionLocationPluginsProvider(IEnumerable<string> extensionLocationPaths)
+        {
+            _extensionLocationPaths = extensionLocationPaths;
+        }
+
+        public IEnumerable<string> GetExtensionLocationPlugins()
+        {
+            return _extensionLocationPaths;
+        }
+
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs:23]: remove extra blank line

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs (hunk 3 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs:26]: > is 'j' the right parameter? what other options do we have?

I changed to extensionsPaths . After that  already we have e parameter for engine . Can I use  'p' ?

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 7 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs]: removed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 8 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs]: removed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs (hunk 1 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs:43]: I added it . Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs (hunk 2 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen.Tests/UnitTests/Providers/DomainModelDefinitionProvidersProvider.cs:49]: I added it . Thanks

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs (hunk 4 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs]: Fixed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs (hunk 5 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs]: Fixed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs (hunk 6 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs]: Fixed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 9 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs]: Fixed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 10 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs]: removed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs (hunk 11 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs]: removed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs [hunk 12]
```diff
@@ -21,5 +22,8 @@ public class Options
 
         [Option('i', "includePlugins", Required = false, HelpText = "Flag to determine if plugin assemblies are included", Default = false)]
         public bool IncludePlugins { get; set; }
+
+        [Option('j', "extensionsPaths", Required = false, HelpText = "Array of paths for the Extension location to determine if plugin assemblies are included")]
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs:26]: lets remove the short name for the parameter and just use `--extensionsPaths` explicitly

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 13]
```diff
@@ -22,10 +22,11 @@ public class DomainModelDefinitionProvidersProvider : IDomainModelDefinitionsPro
 
         private readonly string _solutionPath;
         private readonly ILog Logger = LogManager.GetLogger(typeof(DomainModelDefinitionProvidersProvider));
+        private readonly IExtensionLocationPluginsProvider _extensionLocationPluginsProviderProvider;
         private readonly IIncludePluginsProvider _includePluginsProvider;
         private readonly string _extensionsPath;
 
-        public DomainModelDefinitionProvidersProvider(ICodeRepositoryProvider codeRepositoryProvider, IIncludePluginsProvider includePluginsProvider)
+        public DomainModelDefinitionProvidersProvider(ICodeRepositoryProvider codeRepositoryProvider, IExtensionLocationPluginsProvider extensionLocationPluginsProviderProvider, IIncludePluginsProvider includePluginsProvider)
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:29]: This formatting needs fixing please

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 14]
```diff
@@ -62,28 +65,44 @@ public IEnumerable<IDomainModelDefinitionsProvider> DomainModelDefinitionProvide
...
+                    directoriesToEvaluate = directoriesToEvaluate
+                        .Concat(GetProjectDirectoriesToEvaluate(extensionPath))
                         .ToArray();
+                }
+                else
+                {
+                    throw new Exception(
+                        $"Unable to find extension Location project path  at location {extensionPath}.");
+                }
+                directoriesToEvaluate = directoriesToEvaluate
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:93]: add blank line please

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 15]
```diff
@@ -62,28 +65,44 @@ public IEnumerable<IDomainModelDefinitionsProvider> DomainModelDefinitionProvide
...
                 CodeRepositoryConventions.EdFiOdsFolderName);
 
-            if (_includePluginsProvider.IncludePlugins() && Directory.Exists(_extensionsPath))
+            directoriesToEvaluate = GetProjectDirectoriesToEvaluate(edFiOdsImplementationApplicationPath)
+                .Concat(GetProjectDirectoriesToEvaluate(edFiOdsApplicationPath))
+                .ToArray();
+
+            var extensionPaths = _extensionLocationPluginsProviderProvider.GetExtensionLocationPlugins();
+
+            foreach (var extensionPath in extensionPaths)
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:80]: Can this be converted to a linq statement?

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    public class ExtensionLocationPluginsProvider : IExtensionLocationPluginsProvider
+    {
+        private readonly IEnumerable<string> _extensionPaths;
+
+        public ExtensionLocationPluginsProvider(IEnumerable<string> extensionPaths)
+        {
+            _extensionPaths = extensionPaths;
+        }
+
+        public IEnumerable<string> GetExtensionLocationPlugins()
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs:19]: How often is this called? If more than once, we should return an array instead

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+using System.Collections.Generic;
+
+namespace EdFi.Ods.CodeGen.Providers.Impl
+{
+    public class ExtensionLocationPluginsProvider : IExtensionLocationPluginsProvider
+    {
+        private readonly IEnumerable<string> _extensionPaths;
+
+        public ExtensionLocationPluginsProvider(IEnumerable<string> extensionPaths)
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs:14]: should we be using an immutable parameter?

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 15 above)
Comment (jamessmckay) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:80]: Also we should have a test that validates this logic please

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs [hunk 18]
```diff
@@ -70,6 +73,23 @@ public IEnumerable<AssemblyData> Get()
                 .Select(Create)
                 .ToList();
 
+            var extensionPaths = _extensionLocationPluginsProviderProvider.GetExtensionLocationPlugins();
+
+            foreach (var extensionPath in extensionPaths)
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:78]: can this be  linq statement? also we should have a test that validates this logic

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs [hunk 19]
```diff
@@ -21,5 +22,8 @@ public class Options
...
+        public IEnumerable<string> ExtensionsPaths { get; set; }
```
Comment (jamessmckay) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs:27]: Should this be an immutable object?

**Review by jamessmckay — CHANGES_REQUESTED** [review]

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs (hunk 12 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs]: yes. Fixed it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 14 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs]: Added it

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs (hunk 16 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs]: 2 times .  I changed to string array now

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs (hunk 17 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/ExtensionLocationPluginsProvider.cs]: I changed to string array now

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs (hunk 19 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Options.cs]: it should be IEnumerable<string> for any array with CommandLine

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs (hunk 15 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs]: Added unit test method

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs (hunk 18 above)
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:78]: Added unit test method

**Review by tbmeador-dlp — CHANGES_REQUESTED** [review]

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs [hunk 20]
```diff
@@ -158,6 +178,8 @@ public IEnumerable<AssemblyData> Get()
                 assemblyDatas.Add(legacyAssemblyData);
             }
 
+            assemblyDatas = assemblyDatas.GroupBy(x => x.AssemblyName).Select(x => x.First()).ToList();
```
[the remark this answers is not in the document]
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:181]: As Discussed , I removed this code and add throw condition to handle -IncludePlugins --ExtensionPaths C:\\OSS-Workspace\\Ed-Fi-Extensions\\Extensions\\

[the remark this answers is not in the document]
Comment (semalaiappan) [reply Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/AssemblyDataProvider.cs:181]: throw new Exception(
                        $"Unable to find model definitions file for extensions project {modelProject.Name} at location {metadataFile.FullName}.");

File: Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs [hunk 21]
```diff
@@ -103,9 +122,14 @@ public IEnumerable<IDomainModelDefinitionsProvider> DomainModelDefinitionProvide
                         $"Unable to find model definitions file for extensions project {modelProject.Name} at location {metadataFile.FullName}.");
                 }
 
+                if (domainModelDefinitionsByPath.ContainsKey(modelProject.Name))
+                {
+                    throw new Exception($"Cannot process two same extensions project {modelProject.Name}.");
```
Comment (tbmeador-dlp) [inline Utilities/CodeGeneration/EdFi.Ods.CodeGen/Providers/Impl/DomainModelDefinitionProvidersProvider.cs:127]:
```suggestion
                    throw new Exception($"Found duplicate extension projects for '{modelProject.Name}'.");
```

**Review by tbmeador-dlp — APPROVED** [review]
